src/CakeSlayer/Password.hs

```diff
@@ -0,0 +1,74 @@
+-- | User passwords. Uses @bcrypt@ password hashing.
+
+module CakeSlayer.Password
+       ( PasswordHash (unPasswordHash)
+       , PasswordPlainText (..)
+       , unsafePwdHash
+       , mkPasswordHashWithPolicy
+       , mkPasswordHash
+       , verifyPassword
+       , generateRandomPassword
+       ) where
+
+import Data.Aeson (FromJSON, ToJSON)
+import Database.PostgreSQL.Simple.FromField (FromField)
+import Database.PostgreSQL.Simple.ToField (ToField)
+import Elm (Elm)
+
+import CakeSlayer.Error (WithError, throwOnNothingM)
+import CakeSlayer.Random (mkRandomString)
+
+import qualified Crypto.BCrypt as BC
+
+
+-- | Password hash.
+newtype PasswordHash = PasswordHash
+    { unPasswordHash :: Text
+    } deriving stock (Show, Generic)
+      deriving newtype (Eq, FromField, ToField, FromJSON, ToJSON, Elm)
+
+-- | Unsafe function for constructing 'PasswordHash'. Used mostly for testing.
+unsafePwdHash :: Text -> PasswordHash
+unsafePwdHash = PasswordHash
+
+-- | Password in plain text.
+newtype PasswordPlainText = PasswordPlainText
+    { unPasswordPlainText :: Text
+    } deriving stock (Show, Generic)
+      deriving newtype (Eq, FromJSON, ToJSON, Elm)
+
+
+-- | Generates a password hash given the hashing policy and its plane text.
+-- This has to be done in IO asy generating the salt requires RNG.
+mkPasswordHashWithPolicy
+    :: forall m err . (WithError err m, MonadIO m)
+    => err
+    -> BC.HashingPolicy
```

Comment:
What do you think if this function will return `Maybe` instead of throwing given error?
And then users can use `throwOnNothingM` when they use this function? There're two reasons for this:

1. It will make code here simpler.
2. I'm not sure that `throwOnNothingM` will report proper source code position... In this case it could report line from this library module which is not very useful for user. `CallStack`'s are very fragile and inconvenient :disappointed: Maybe in some future we can specify how deep `CallStack` you want to display.